ocaml/xapi/xapi_vm_group_helpers.ml

```diff
@@ -0,0 +1,173 @@
+      true
+    with Not_found -> false
+  in
+  let group_uuid = Db.VM_group.get_uuid ~__context ~self:group in
+  records
+  |> List.filter (fun (_, record) ->
+         record.API.message_name
+         = fst Api_messages.breach_on_vm_anti_affinity_rules
+         && contains_substring record.API.message_body group_uuid
+     )
+
+(** If it is breach and no alerts existed, generate one, 
+    If it is not breach and alerts existed, dismiss the existed alert *)
+let update_vm_placement_alert_with_group ~__context ~group ~records =
+  try
+    let breach_on_host = check_breach ~__context ~group in
+    let alert_records = filter_alert_with_group ~__context ~group ~records in
+    debug "The length of group (UUID: %s) is: %d"
+      (Db.VM_group.get_uuid ~__context ~self:group)
+      (List.length alert_records) ;
+    match (alert_records, breach_on_host) with
+    | [], Some host ->
+        report_anti_affinity_alert ~__context ~group ~host
+    | records, None ->
+        List.iter
+          (fun (ref, _) ->
+            Helpers.call_api_functions ~__context (fun rpc session_id ->
+                Client.Client.Message.destroy ~rpc ~session_id ~self:ref
+            )
+          )
+          records
+    | _, _ ->
+        ()
+  with _ -> warn "Update VM anti-affinity alert failed"
+
+let update_vm_placement_alert_with_vm ~__context ~vm =
+  let group_opt =
+    List.find_opt
+      (fun g -> Db.VM_group.get_placement ~__context ~self:g = `anti_affinity)
+      (Db.VM.get_groups ~__context ~self:vm)
+  in
+  match group_opt with
+  | None ->
+      ()
+  | Some group ->
+      update_vm_placement_alert_with_group ~__context ~group
+        ~records:(get_all_records ~__context)
+
+let remove_vm_placement_alert_with_group ~__context ~group ~records =
+  filter_alert_with_group ~__context ~group ~records
+  |> List.iter (fun (ref, _) ->
+         Helpers.call_api_functions ~__context (fun rpc session_id ->
+             Client.Client.Message.destroy ~rpc ~session_id ~self:ref
+         )
+     )
+
+let update_vm_placement_alert ~__context ~groups =
+  let records = get_all_records ~__context in
+  groups
+  |> List.filter (fun g ->
+         Db.VM_group.get_placement ~__context ~self:g = `anti_affinity
+     )
+  |> List.iter (fun group ->
+         update_vm_placement_alert_with_group ~__context ~group ~records
+     )
+
+let remove_vm_placement_alert ~__context ~groups =
+  let records = get_all_records ~__context in
+  groups
+  |> List.filter (fun g ->
+         Db.VM_group.get_placement ~__context ~self:g = `anti_affinity
+     )
+  |> List.iter (fun group ->
+         remove_vm_placement_alert_with_group ~__context ~group ~records
+     )
+
+let update_vm_placement_alert_on_features_change ~__context ~old_restrictions
+    ~new_restrictions =
+  try
+    let feature_name = Features.name_of_feature Features.VM_anti_affinity in
+    let feature = "restrict_" ^ String.lowercase_ascii feature_name in
```

Comment:
This should be obtained by `fst Features.(string_of_feature VM_anti_affinity)`

Reply:
According to:https://github.com/xapi-project/xen-api/blob/4752accf17adfe0323850b5276e79496bf78feb5/ocaml/xapi-types/features.mli . I think the string_of_feature is not an exposed interface

Reply:
OK. Then `Pool_features.is_enabled` can be an example to determine if a feature is enabled.

Reply:
`Pool_features.is_enabled` can only get the new restriction state, we still need something like above to get the old restriction state.
and I guess it may introduce the dependency cycle as well

Reply:
You can expose the necessary interface to handle this.
`let feature = "restrict_" ^ String.lowercase_ascii feature_name in` can't be accepted.

Reply:
Great. then I expose the `string_of_feature` to get the restriction of the feature with latest implementation